core/src/main/java/io/grpc/Contexts.java

```diff
@@ -0,0 +1,239 @@
+     */
+    Context apply(Context toTransform, MethodDescriptor<?, ?> method, Metadata headers);
+  }
+
+  /**
+   * Transformer for client calls.
+   */
+  public static interface ClientContextTransformer {
+    /**
+     * Apply a transformation to the given {@link Context} for a dispatched call.
+     * @param toTransform the context to transform.
+     * @param method the method that is being called by the client.
+     * @param options the runtime options for the call
+     * @return a new context.
+     */
+    Context apply(Context toTransform, MethodDescriptor<?, ?> method, CallOptions options);
+  }
+
+  /**
+   * Apply a {@link ClientContextTransformer} for calls that pass through this interceptor.
+   */
+  private static class ContextAugmentingClientInterceptor implements ClientInterceptor {
+
+    private final ClientContextTransformer transformer;
+
+    private ContextAugmentingClientInterceptor(ClientContextTransformer transformer) {
+      this.transformer = transformer;
+    }
+
+    @Override
+    public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(
+        MethodDescriptor<ReqT, RespT> method,
+        CallOptions callOptions,
+        Channel next) {
+      Context newContext = transformer.apply(Context.current(), method, callOptions);
+      Context previous = newContext.attach();
+      try {
+        ClientCall<ReqT, RespT> call = next.newCall(method, callOptions);
+        return new ForwardingClientCall.SimpleForwardingClientCall<ReqT, RespT>(call) {
+          @Override
+          public void start(Listener<RespT> responseListener, Metadata headers) {
+            super.start(responseListener, headers);
+          }
+        };
+      } finally {
+        previous.attach();
+      }
+    }
+  }
+
+  /**
+   * Apply a {@link ServerContextTransformer} to calls that pass through this interceptor.
+   */
+  private static class ContextAugmentingServerInterceptor implements ServerInterceptor {
+
+    private final ServerContextTransformer transformer;
+
+    private ContextAugmentingServerInterceptor(ServerContextTransformer transformer) {
+      this.transformer = transformer;
+    }
+
+    @Override
+    public <ReqT, RespT> ServerCall.Listener<ReqT> interceptCall(
+        MethodDescriptor<ReqT, RespT> method, ServerCall<RespT> call,
+        Metadata headers, ServerCallHandler<ReqT, RespT> next) {
+      Context newContext = transformer.apply(Context.current(), method, headers);
+      Context previous = newContext.attach();
+      try {
+        return new ContextualizedServerCallListener<ReqT>(next.startCall(method, call, headers),
+            newContext);
+      } finally {
+        previous.attach();
+      }
+    }
+  }
+
+  /**
+   * Implementation of {@link io.grpc.ForwardingClientCallListener} that attaches a context before
+   * dispatching calls to the delegate and detaches them after the call completes.
+   */
+  public static class ContextualizedClientCallListener<RespT>
```

Comment:
I was asking more about the usage pattern. I would expect just to set the context when calling `call.start()`, and it will automatically be available on the listener.